typeclasses/maps.py

```diff
@@ -0,0 +1,252 @@
+# -*- encoding: utf-8 -*-
+
+from items import Item
+from utils import utils
+
+TILES = {
+    'WATER': u'≈≈≈',
+    'DEEPWATER': u'≋≋≋',
+    'VEGETATION': u'~↟~',
+    'GRASS': u' . ',
+}
+CELL_WIDTH = 3
+PADDING = 1
+
+class Map(Item):
+    '''
+    The Map typeclass creates an empty map which fills up
+    as its owner travels through Ainneve.
+    When looking at the map, a representation of the world
+    as explored so far is shown.
+
+    Rooms can customize the map tile drawn for them in several ways:
+        - by specifying a `map_tile()` function that takes the requesting map
+          and returns a three-character string to draw on the map
+        - by specifying a `map_tile` property containing
+          a three-character string to draw on the map
+        - by specifying a `map_tile` Attribute containing
+          a three-character string to draw on the map
+        - by specifying a `terrain` present in the TILES dictionary in maps.py
+    If no tile can be found, the map will display `?!?` as the tile.
+    '''
+    weight = 1.0
+
+    def at_object_creation(self):
+        super(Map, self).at_object_creation()
+        self.db.map_data = {}
+        self.db.x = self.db.y = 0
+        self.db.off_map = False
+        self.map_current_room()
+
+    @property
+    def room(self):
+        '''
+        Finds the room that this object is currently in,
+        or None if it isn’t in a room.
+        This still works if the map is inside one or more nested objects.
+        '''
+        loc = self.location
+        if loc is None:
+            return loc
+
+        done = False
+        while 1:
+            if loc.location is not None:
+                loc = loc.location
+            else:
+                break
+        return loc
+
+    @property
+    def current_character(self):
+        '''
+        This returns the character carrying the map,
+        or None if the map isn’t currently being carried by a character.
+        '''
+        character = self.location
+        if character is None:
+            return character
+
+        done = False
+        while 1:
+            if character is None:
+                return None
+            if not character.is_typeclass('characters.Character'):
+                character = character.location
+            else:
+                break
```

Comment:
Same thing here, I might suggest a way to break out of the loop after some amount of iterations. I don't think this is as likely to cause problems though.